File: bindings/node/facade/routing.ts
/**
 * Ergonomic facade over the generated routing binding.
 *
 * Flooding always works but costs every node airtime and power on every packet.
 * Once a mesh has settled, most traffic goes to a few known places, and a node
 * that remembers the way can forward to one neighbour instead of shouting at the
 * whole network. Routing is that optimisation, and it falls back to flooding
 * rather than failing whenever it does not know the way.
 *
 * The routing action is re-exported as a runtime {@link ForwardAction} object,
 * because the generated enum is types-only.
 *
 * @packageDocumentation
 */

import { ROUTING_TABLE_CAPACITY, Router as NativeRouter } from '../index'

/** The number of routes a routing table holds. */
export const TABLE_CAPACITY = ROUTING_TABLE_CAPACITY

/**
 * What to do with a packet bound for a given node.
 *
 * Provided as a runtime object plus a matching string-union type so it works as
 * both a value (`ForwardAction.Relay`) and a type annotation.
 */
export const ForwardAction = {
  /** The packet is for this node; hand it to the application. */
  Deliver: 'Deliver',
  /** A route is known; unicast the packet to the next hop reported alongside. */
  Relay: 'Relay',
  /** No route is known; fall back to flooding the packet. */
  Flood: 'Flood',
} as const

/** One of the {@link ForwardAction} choices. */
export type ForwardAction = (typeof ForwardAction)[keyof typeof ForwardAction]

/** A routing decision, and the neighbour it names when there is one. */
export interface ForwardDecision {
  /** What to do with the packet. */
  action: ForwardAction
  /** The neighbour to unicast to, or `null` unless the action is `Relay`. */
  nextHop: number | null
}

/** A learned way to reach one node. */
export interface Route {
  /** The node this route reaches. */
  dst: number
  /** The neighbour to send a packet to on the way there. */
  nextHop: number
  /** What the route costs, usually in hops. */
  cost: number
}

/**
 * One node routing table, learned from the traffic the node hears.
 *
 * The core table is generic over its size, which cannot cross the native
 * boundary, so this is built at one documented size: {@link TABLE_CAPACITY}.
 */
export class Router {
  readonly #inner: NativeRouter

  /**
   * Creates an empty routing table for a node.
   *
   * @param address - The address of this node, which is what a routing decision
   *   recognises as a local delivery.
   */
  constructor(address: number) {
    this.#inner = new NativeRouter(address)
  }

  /** The address this router answers for. */
  get address(): number {
    return this.#inner.address
  }

  /** How many routes the table currently holds. */
  get size(): number {
    return this.#inner.len
  }

  /** Whether the table has learned nothing yet. */
  get isEmpty(): boolean {
    return this.#inner.isEmpty
  }

  /** How many routes the table can hold. */
  get capacity(): number {
    return this.#inner.capacity
  }

  /**
   * Learns a route from a packet that arrived.
   *
   * @param origin - The node the packet came from.
   * @param via - The neighbour it arrived through.
   * @param cost - What that path costs, usually a hop count.
   * @returns Whether the table changed. It keeps the cheapest way it knows to
   *   each node, and when full gives up the most expensive route to make room.
   */
  observe(origin: number, via: number, cost: number): boolean {
    return this.#inner.observe(origin, via, cost)
  }

  /**
   * Returns the neighbour on the way to a node.
   *
   * @param dst - The node to reach.
   * @returns The next hop, or `null` when no route is known.
   */
  nextHop(dst: number): number | null {
    return this.#inner.nextHop(dst) ?? null
  }

  /**
   * Returns what the known route to a node costs.
   *
   * @param dst - The node to reach.
   * @returns The cost, or `null` when no route is known.
   */
  cost(dst: number): number | null {
    return this.#inner.cost(dst) ?? null
  }

  /**
   * Returns the whole route to a node.
   *
   * @param dst - The node to reach.
   * @returns The route, or `null` when no route is known.
   */
  route(dst: number): Route | null {
    return this.#inner.route(dst) ?? null
  }

  /**
   * Decides what to do with a packet bound for a node.
   *
   * @param dst - The node the packet is addressed to.
   * @returns The decision, carrying a next hop only when it says to relay.
   */
  forward(dst: number): ForwardDecision {
    // The generated object leaves an absent next hop undefined; null says the
    // same thing in the shape the rest of this package uses.
    const decision = this.#inner.forward(dst)
    return { action: decision.action, nextHop: decision.nextHop ?? null }
  }

  /**
   * Forgets the route to a node, for example after it stops answering.
   *
   * @param dst - The node to forget.
   */
  forget(dst: number): void {
    this.#inner.forget(dst)
  }
}

/**
 * Creates an empty routing table for a node.
 *
 * @param address - The address of this node, which is what a routing decision
 *   recognises as a local delivery.
 * @returns The routing table, ready to learn from the traffic the node hears.
 */
export function router(address: number): Router {
  return new Router(address)
}

File: bindings/dotnet/src/Pamoja.Routing/Routing.cs
using Pamoja.Core;
using Pamoja.Native.Interop;

namespace Pamoja.Routing;

/// <summary>What to do with a packet bound for a given node.</summary>
public enum ForwardAction
{
    /// <summary>The packet is for this node; hand it to the application.</summary>
    Deliver = 0,

    /// <summary>A route is known; unicast the packet to the next hop reported alongside.</summary>
    Relay = 1,

    /// <summary>No route is known; fall back to flooding the packet.</summary>
    Flood = 2,
}

/// <summary>A routing decision, and the neighbour it names when there is one.</summary>
/// <param name="Action">What to do with the packet.</param>
/// <param name="NextHop">
/// The neighbour to unicast to, or <c>null</c> unless the action is
/// <see cref="ForwardAction.Relay"/>.
/// </param>
public readonly record struct ForwardDecision(ForwardAction Action, uint? NextHop);

/// <summary>A learned way to reach one node.</summary>
/// <param name="Dst">The node this route reaches.</param>
/// <param name="NextHop">The neighbour to send a packet to on the way there.</param>
/// <param name="Cost">What the route costs, usually in hops.</param>
public readonly record struct Route(uint Dst, uint NextHop, ushort Cost);

/// <summary>One node routing table, learned from the traffic the node hears.</summary>
/// <remarks>
/// Flooding always works but costs every node airtime and power on every packet.
/// A node that remembers the way can forward to one neighbour instead, and falls
/// back to flooding rather than failing whenever it does not know the way. The
/// core table is generic over its size, which cannot cross the C ABI, so this one
/// is sized when it is built.
/// </remarks>
public sealed class Router : IDisposable
{
    private readonly NativeHandle _handle;

    /// <summary>Creates an empty routing table for a node.</summary>
    /// <param name="address">
    /// The address of this node, which is what a routing decision recognises as a
    /// local delivery.
    /// </param>
    /// <param name="capacity">
    /// How many routes to make room for. A capacity of 0 floods every unknown
    /// destination, which is the behaviour with no table at all.
    /// </param>
    /// <exception cref="PamojaException">The native table could not be created.</exception>
    public Router(uint address, int capacity = NativeMethods.RoutingDefaultCapacity)
    {
        _handle = NativeHandle.Create(
            NativeMethods.pamoja_router_new(address, (nuint)capacity),
            NativeMethods.pamoja_router_free,
            "router");
    }

    /// <summary>The address this router answers for.</summary>
    public uint Address => _handle.Use(NativeMethods.pamoja_router_address);

    /// <summary>How many routes the table currently holds.</summary>
    public int Count => checked((int)_handle.Use(NativeMethods.pamoja_router_len));

    /// <summary>How many routes the table can hold.</summary>
    public int Capacity =>
        checked((int)_handle.Use(NativeMethods.pamoja_router_capacity));

    /// <summary>Learns a route from a packet that arrived.</summary>
    /// <param name="origin">The node the packet came from.</param>
    /// <param name="via">The neighbour it arrived through.</param>
    /// <param name="cost">What that path costs, usually a hop count.</param>
    /// <returns>
    /// Whether the table changed. It keeps the cheapest way it knows to each node,
    /// and when full gives up the most expensive route to make room for a cheaper
    /// one.
    /// </returns>
    public bool Observe(uint origin, uint via, ushort cost) =>
        _handle.Use(handle => NativeMethods.pamoja_router_observe(handle, origin, via, cost));

    /// <summary>Returns the neighbour on the way to a node.</summary>
    /// <param name="dst">The node to reach.</param>
    /// <returns>The next hop, or <c>null</c> when no route is known.</returns>
    public uint? NextHop(uint dst) =>
        _handle.Use(handle =>
            NativeMethods.pamoja_router_next_hop(handle, dst, out uint next) ? next : (uint?)null);

    /// <summary>Returns what the known route to a node costs.</summary>
    /// <param name="dst">The node to reach.</param>
    /// <returns>The cost, or <c>null</c> when no route is known.</returns>
    public ushort? Cost(uint dst) =>
        _handle.Use(handle =>
            NativeMethods.pamoja_router_cost(handle, dst, out ushort cost) ? cost : (ushort?)null);

    /// <summary>Returns the whole route to a node.</summary>
    /// <param name="dst">The node to reach.</param>
    /// <returns>The route, or <c>null</c> when no route is known.</returns>
    public Route? RouteTo(uint dst) =>
        _handle.Use(handle =>
            NativeMethods.pamoja_router_route(handle, dst, out PamojaRoute route)
                ? new Route(route.Dst, route.NextHop, route.Cost)
                : (Route?)null);

    /// <summary>Decides what to do with a packet bound for a node.</summary>
    /// <param name="dst">The node the packet is addressed to.</param>
    /// <returns>The decision, carrying a next hop only when it says to relay.</returns>
    public ForwardDecision Forward(uint dst) =>
        _handle.Use(handle =>
        {
            PamojaForward action = NativeMethods.pamoja_router_forward(handle, dst, out uint next);
            return action == PamojaForward.Relay
                ? new ForwardDecision(ForwardAction.Relay, next)
                : new ForwardDecision((ForwardAction)action, null);
        });

    /// <summary>Forgets the route to a node, for example after it stops answering.</summary>
    /// <param name="dst">The node to forget.</param>
    public void Forget(uint dst) =>
        _handle.Use(handle => NativeMethods.pamoja_router_forget(handle, dst));

    /// <inheritdoc/>
    public void Dispose() => _handle.Dispose();
}
